receiver/awsxrayreceiver/internal/translator/cause.go

```diff
@@ -0,0 +1,102 @@
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/tracesegment"
+)
+
+const (
+	newLine   = "\n"
+	separator = ":"
+)
+
+func addCause(seg *tracesegment.Segment, span *pdata.Span) {
+	if seg.Cause == nil {
+		return
+	}
+
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/cause.go#L51
+	// In the `rawExpectedSegmentForInstrumentedApp` X-Ray segment example in
+	// tracesegment_test.go, you can see that sometimes the HTTP Response is not
+	// set but the Cause field is set for the root segment. And the actual HTTP status
+	// can only be found in one of the (nested) subsegmets. So we need to signal that
+	// in this case, the status of the span is not otlptrace.Status_Ok by
+	// temporarily setting the status to otlptrace.Status_UnknownError. This will be
+	// updated to a more specific error in the `segToSpans()` in translator.go once
+	// we traverse through all the subsegments.
+	if span.Status().Code() == pdata.StatusCode(otlptrace.Status_Ok) {
+		// otlptrace.Status_Ok is the default value after span.Status().InitEmpty()
+		// is called
+		span.Status().SetCode(pdata.StatusCode(otlptrace.Status_UnknownError))
+	}
+
+	switch seg.Cause.Type {
+	case tracesegment.CauseTypeExceptionID:
+		// Right now the X-Ray exporter always genearates a new ID:
+		// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/cause.go#L74
+		// https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/exporter/awsxrayexporter/translator/cause.go#L112
+		// so we can only pass this as part of the status message as a fallback mechanism
+		span.Status().SetMessage(*seg.Cause.ExceptionID)
+	case tracesegment.CauseTypeObject:
+		evts := span.Events()
+		// not sure whether there are existing events, so
+		// append new empty events instead
+		exceptionEventStartIndex := evts.Len()
+		evts.Resize(exceptionEventStartIndex + len(seg.Cause.Exceptions))
+
+		for i, excp := range seg.Cause.Exceptions {
+			evt := evts.At(exceptionEventStartIndex + i)
+			evt.InitEmpty()
+			evt.SetName(conventions.AttributeExceptionEventName)
+			attrs := evt.Attributes()
+			attrs.InitEmptyWithCapacity(2)
+			if excp.Type != nil {
+				attrs.UpsertString(conventions.AttributeExceptionType, *excp.Type)
+				attrs.UpsertString(conventions.AttributeExceptionMessage, *excp.Message)
+				stackTrace := convertStackFramesToStackTraceStr(excp.Stack)
+				attrs.UpsertString(conventions.AttributeExceptionStacktrace, stackTrace)
+				// For now X-Ray's exception data model is not fully supported in the OpenTelemetry
+				// spec, so some information is lost here.
+				// For example, the "cause" ,"remote", ... and some fields within each exception
+				// are dropped.
+			}
+		}
+	}
+}
+
+func convertStackFramesToStackTraceStr(stack []tracesegment.StackFrame) string {
```

Comment:
Let's go ahead and construct a Java stack trace (added the parsing code yesterday) instead of a custom stacktrace format. We can use that for the causes too I believe to avoid the ExceptionID issue